### ISSUE TITLE
Improve network performance

### ISSUE THREAD

**Issue by paulie-g** [gleitz/howdoi#233]
We've improved most of what we can improve in terms of start-up time - yes, a few small things are left and I'll do them - but they're vanishingly small in comparison to the run times in the cache miss case. The most impactful thing remaining is speeding up the network stuff.

We know as soon as we run that we will want to connect to google/bing and stackoverflow. Ideally, we want to open a TLS connection to them right at the top (asynchronously) and then hand it to whatever does our http GETs. Unfortunately, doing so with requests requires massive monkey-patching. Requests is great, but we don't need 'great' here - we're just doing a few GETs and need something that'll take our socket to do them over.

I'm proposing pycurl/libcurl. We can hand it a socket, it's fast and it will eventually do QUIC/HTTP3 once the dust settles. libcurl is also ubiquitous. The api is ugly, but that ugliness need only be confined to/hidden in a couple functions (I might even mock the requests api to the extent we use it).

I expect non-negligible speed-up, at least 100ms even on fast connections. There's even, optimistically, a chance I can hit 500ms on my box. Can't tell for sure until I do the work obviously, might do a profiling run though and see if that yields insight.

With all that said, before I get into this, I want to make sure that this is something you would merge.

**Comment by paulie-g** [comment]
Had a quick shufti with httpstat and there's at least 150ms on the table here from what I can see, maybe more, coupled with another 75+ms from eliminating the requests import. I need a yay or nay from @gleitz on this before I proceed though.

**Comment by gleitz** [comment]
As long as we maintain portability, I'm fine ditching `requests`. Pretty soon the results will come back before you've thought of the question.

**Comment by paulie-g** [comment]
I have a few ideas in that direction ;) My goal is 500ms on a low-latency link with the answer page in SO's varnish cache. May not be doable (at ~800ms currently), but it's something to shoot for. I think 500-600ms would be good user experience.

libcurl/pycurl is about as portable as any binary dependency is going to be. It's got binary wheels for win32/win64 (where howdoi was unlikely to work before I committed the use of appdirs for cache location anyway) and who doesn't have libcurl on anything else? OSX used to have some issues with install, but those issues are long gone as far as I can tell and, frankly, if someone wants to feed the fruit monster that's on them and they've got to be prepared to jump through some hoops ;)

Having said all that, what I will try to do - as long as it's practicable - is to provide a fallback using requests. We have a very narrow exposure to its api, so it shouldn't be too much extra work.

**Comment by paulie-g** [comment]
As an FYI, someone ran benchmarks of curl vs requests here some time ago: https://github.com/svanoort/python-client-benchmarks
TLDR; curl is much, much faster.

For our purposes, it also matters that it is thread-safe and afaict doesn't hold GIL.

**Comment by gleitz** [comment]
Closing for now, but please re-open if you have a PR
